library/std/src/path.rs

```diff
@@ -2686,7 +2694,8 @@ impl fmt::Display for Display<'_> {
 impl cmp::PartialEq for Path {
     #[inline]
     fn eq(&self, other: &Path) -> bool {
-        self.components().eq(other.components())
+        self.components() == other.components()
+            && self.ends_with_separator() == other.ends_with_separator()
```

Comment:
Shouldn't this be implemented on the `Components` level instead of `Path`? Otherwise they'll give different results.

Reply:
Giving different results is correct. The path components are the same i.e. the `Components` iterator would produce the identical sequence of `Component` enum values, even if one path ends in separator and the other does not.

Reply:
If we say that a trailing slash is semantically different then shouldn't components preserve that semantic difference? E.g. the last component could include the trailing slash. Or there could be a final zero-length component. Ideally this would be expressed through a new enum variant but it's exhaustive so we can't extend it 🙁.

Is `assert_eq!(foo, foo.components().as_path())` always true?

Reply:
> I don't think it follows that we need components to preserve that semantic difference.

Well... the thing is that `Components` implements `AsRef<Path>`, so if you have a generic function taking it via `AsRef` then it'll behave differently depending on whether you pass in a `Components` or something else that implements `AsRef<Path>`.

It's consistent with existing documentation, it's clearly spelled out that `path.components()` will strip trailing slashes, but previously that wasn't considered a semantic difference (since `Eq` considered them equal), now it is.